server/webapp/WEB-INF/rails.new/app/controllers/admin/materials/pluggable_scm_controller.rb

```diff
@@ -0,0 +1,212 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+##########################GO-LICENSE-END##################################
+
+module Admin::Materials
+  class PluggableScmController < AdminController
+    include PipelineConfigLoader
+
+    load_pipeline_for_all_actions
+
+    def show_existing
+      assert_load :material, PluggableSCMMaterialConfig.new
+      assert_load :scms, @cruise_config.getSCMs()
+      render layout: false
+    end
+
+    def choose_existing
+      assert_load :material, PluggableSCMMaterialConfig.new
+
+      create_failure_handler = proc do |result, all_errors|
+        @errors = flatten_all_errors(all_errors)
+        @scms = @cruise_config.getSCMs()
+        render :template => "/admin/materials/pluggable_scm/show_existing", :status => result.httpCode(), :layout => false
+      end
+
+      save_popup(params[:config_md5], get_choose_existing_command, create_failure_handler, {:controller => '/admin/materials', :stage_parent => "pipelines", :current_tab => params[:current_tab]}) do
+        assert_load :pipeline, @node.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name]))
+        assert_load :material, @subject
+      end
+    end
+
+    def new
+      scm = com.thoughtworks.go.domain.scm.SCM.new
+      scm.setPluginConfiguration(PluginConfiguration.new(params[:plugin_id], nil))
+      pluggable_scm = PluggableSCMMaterialConfig.new
+      pluggable_scm.setSCMConfig(scm)
+
+      assert_load :material, pluggable_scm
+      assert_load :meta_data_store, meta_data_store
+      render layout: false
+    end
+
+    def create
+      scm = com.thoughtworks.go.domain.scm.SCM.new
+      scm.setPluginConfiguration(PluginConfiguration.new(params[:plugin_id], '1'))
+      pluggable_scm = PluggableSCMMaterialConfig.new
+      pluggable_scm.setSCMConfig(scm)
+      assert_load :material, pluggable_scm
+
+      create_failure_handler = proc do |result, all_errors|
+        @errors = flatten_all_errors(all_errors)
+        @meta_data_store = meta_data_store
+        render :template => "/admin/materials/pluggable_scm/new", :status => result.httpCode(), :layout => false
+      end
+
+      save_popup(params[:config_md5], get_create_command, create_failure_handler, {:controller => '/admin/materials', :stage_parent => "pipelines", :current_tab => params[:current_tab]}) do
+        assert_load :pipeline, @node.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name]))
+        assert_load :material, @subject
+      end
+    end
+
+    def edit
+      assert_load :material, @pipeline.materialConfigs().getByFingerPrint(params[:finger_print])
+      assert_load :meta_data_store, meta_data_store
+      render layout: false
+    end
+
+    def update
+      material = @pipeline.materialConfigs().getByFingerPrint(params[:finger_print])
```

Comment:
Line is too long. [84/80]<br>Do not use parentheses for method calls with no arguments.